include/GafferCycles/IECoreCyclesPreview/IECoreCycles.h

```diff
@@ -64,4 +61,19 @@ IECORECYCLES_API int patchVersion();
 /// Returns a string of the form "major.minor.patch"
 IECORECYCLES_API const std::string &versionString();
 
+/// Returns all device data (for python bindings)
+IECORECYCLES_API const IECore::CompoundDataPtr devices();
```

Comment:
Should be returning `const IECore::CompoundData *` here I think. The caller doesn't need ownership because the data is owned statically, so a raw pointer is fine. But the caller shouldn't be allowed to modify the data, hence it should be const. Note that `const CompoundDataPtr` is a pointer to a mutable object - the `const` applies to the pointer itself in this case (so you couldn't re-assign to it).

Once we've done this, we'll either need to take a copy or use a const-cast in the Python bindings (since Python has no concept of const). Since it's just a one-off cost, I'd suggest taking a copy.

Reply:
Thanks! Rebased to fix the merge conflict and squashed in returning `const IECore::CompoundData *`  and taking a copy in the Python bindings as they were minor changes. Everything else is as it was...